File: soc-server/internal/services/rule_service.go
// ==============================================================================
// Package services - Rule Service
// File: internal/services/rule_service.go
// Mô tả: Business logic cho Detection Rules management.
//         Quản lý Rule trong database và gọi trigger hot-reload Rule Engine.
// ==============================================================================

package services

import (
	"encoding/json"
	"fmt"
	"log"
	"time"

	"soc-server/internal/models"
	"soc-server/internal/rules"

	"gorm.io/gorm"
)

// DefaultRules returns a set of starter SOC detection rules that are useful for demo and admin usage.
func DefaultRules() []models.Rule {
	return []models.Rule{
		{
			ID:               "RULE-POWERSHELL-ENCODED",
			Name:             "PowerShell Encoded Command Execution",
			Severity:         "high",
			EventType:        "process_creation",
			Conditions:       `[{"field":"process_name","operator":"equals","value":"powershell.exe"},{"field":"command_line","operator":"contains","value":"-EncodedCommand"}]`,
			MITRETactic:      "Execution",
			MITRETechniqueID: "T1059.001",
			Description:      "Detects PowerShell commands executed with encoded payloads, often used for script obfuscation.",
			IsActive:         true,
			Source:           "database",
		},
		{
			ID:               "RULE-RANSOMWARE-VSS",
			Name:             "Ransomware VSS Shadow Delete Activity",
			Severity:         "critical",
			EventType:        "command_execution",
			Conditions:       `[{"field":"process_name","operator":"in","value":"vssadmin.exe"},{"field":"command_line","operator":"contains","value":"shadow delete"}]`,
			MITRETactic:      "Impact",
			MITRETechniqueID: "T1490",
			Description:      "Flags attempts to delete volume shadow copies, a common ransomware behavior.",
			IsActive:         true,
			Source:           "database",
		},
		{
			ID:               "RULE-NETWORK-C2",
			Name:             "Outbound C2 Communication Pattern",
			Severity:         "medium",
			EventType:        "network_connection",
			Conditions:       `[{"field":"dst_ip","operator":"not_equals","value":"10.0.0.0/8"},{"field":"protocol","operator":"equals","value":"tcp"}]`,
			MITRETactic:      "Command and Control",
			MITRETechniqueID: "T1071",
			Description:      "Looks for suspicious outbound connections outside the internal network segmentation.",
			IsActive:         true,
			Source:           "database",
		},
	}
}

// RuleService - Service xử lý business logic cho Detection Rules
type RuleService struct {
	db         *gorm.DB
	ruleEngine *rules.RuleEngine
}

// NewRuleService - Khởi tạo RuleService
func NewRuleService(db *gorm.DB, ruleEngine *rules.RuleEngine) *RuleService {
	return &RuleService{
		db:         db,
		ruleEngine: ruleEngine,
	}
}

// GetAllRules - Lấy toàn bộ danh sách rules từ Database
func (s *RuleService) GetAllRules() ([]models.Rule, error) {
	var dbRules []models.Rule
	err := s.db.Order("created_at DESC").Find(&dbRules).Error
	return dbRules, err
}

// CreateRule - Thêm Rule mới vào database và Hot-reload Engine
func (s *RuleService) CreateRule(rule *models.Rule) error {
	if rule.Source == "" {
		rule.Source = "database"
	}
	if rule.ID == "" {
		rule.ID = fmt.Sprintf("RULE-%d", time.Now().UnixNano())
	}

	// Validate JSONB conditions
	if !isValidJSON(rule.Conditions) {
		return fmt.Errorf("conditions không phải là JSON hợp lệ")
	}

	if err := s.db.Create(rule).Error; err != nil {
		return err
	}

	log.Printf("[RULE SERVICE] Đã tạo Rule mới trong DB: '%s'. Kích hoạt Hot-reload...", rule.ID)
	return s.ruleEngine.ReloadRules()
}

// ToggleRule - Bật/tắt 1 Rule (IsActive) trong DB và Hot-reload
func (s *RuleService) ToggleRule(id string, isActive bool) error {
	result := s.db.Model(&models.Rule{}).Where("id = ?", id).Update("is_active", isActive)
	if result.RowsAffected == 0 {
		return fmt.Errorf("không tìm thấy Rule '%s' trong Database", id)
	}
	if result.Error != nil {
		return result.Error
	}

	log.Printf("[RULE SERVICE] Rule '%s' chuyển trạng thái active=%v. Kích hoạt Hot-reload...", id, isActive)
	return s.ruleEngine.ReloadRules()
}

// UpdateRule - Cập nhật Rule trong DB (nếu chưa có trong DB sẽ tự động tạo mới)
func (s *RuleService) UpdateRule(id string, rule *models.Rule) error {
	if rule == nil {
		return fmt.Errorf("rule không được null")
	}
	if !isValidJSON(rule.Conditions) {
		return fmt.Errorf("conditions không phải là JSON hợp lệ")
	}
	if rule.ID == "" {
		rule.ID = id
	}

	var existing models.Rule
	err := s.db.Where("id = ?", id).First(&existing).Error
	if err != nil {
		// Nếu chưa có trong DB -> tạo mới
		if rule.Source == "" {
			rule.Source = "database"
		}
		if err := s.db.Create(rule).Error; err != nil {
			return err
		}
	} else {
		// Đã có -> Cập nhật
		updates := map[string]interface{}{
			"name":               rule.Name,
			"severity":           rule.Severity,
			"event_type":         rule.EventType,
			"conditions":         rule.Conditions,
			"mitre_tactic":       rule.MITRETactic,
			"mitre_technique_id": rule.MITRETechniqueID,
			"description":        rule.Description,
			"is_active":          rule.IsActive,
		}
		if rule.Source != "" {
			updates["source"] = rule.Source
		}
		if err := s.db.Model(&existing).Updates(updates).Error; err != nil {
			return err
		}
	}

	log.Printf("[RULE SERVICE] Đã cập nhật Rule '%s'. Kích hoạt Hot-reload...", id)
	return s.ruleEngine.ReloadRules()
}

func (s *RuleService) DeleteRule(id string) error {
	result := s.db.Where("id = ?", id).Delete(&models.Rule{})
	if result.RowsAffected == 0 {
		return fmt.Errorf("không tìm thấy Rule '%s' trong DB", id)
	}
	if result.Error != nil {
		return result.Error
	}
	return s.ruleEngine.ReloadRules()
}

// Helper kiểm tra JSON hợp lệ
func isValidJSON(s string) bool {
	var js json.RawMessage
	return json.Unmarshal([]byte(s), &js) == nil
}

File: soc-server-python/app/services/rule_service.py
# ==============================================================================
# app/services/rule_service.py - Detection Rule Service
# Tương đương: internal/services/rule_service.go
# ==============================================================================

import json
import logging
import time
from typing import List, Optional, Dict, Any
from datetime import datetime

from sqlalchemy import select, delete, desc
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.rule import Rule
from app.rules.engine import global_rule_engine, RuleEngine

logger = logging.getLogger(__name__)


def default_rules() -> List[dict]:
    return [
        {
            "id": "RULE-POWERSHELL-ENCODED",
            "name": "PowerShell Encoded Command Execution",
            "severity": "high",
            "event_type": "process_creation",
            "conditions": json.dumps([
                {"field": "process_name", "operator": "equals", "value": "powershell.exe"},
                {"field": "command_line", "operator": "contains", "value": "-EncodedCommand"}
            ]),
            "mitre_tactic": "Execution",
            "mitre_technique_id": "T1059.001",
            "description": "Detects PowerShell commands executed with encoded payloads, often used for script obfuscation.",
            "is_active": True,
            "source": "database",
        },
        {
            "id": "RULE-RANSOMWARE-VSS",
            "name": "Ransomware VSS Shadow Delete Activity",
            "severity": "critical",
            "event_type": "command_execution",
            "conditions": json.dumps([
                {"field": "process_name", "operator": "in", "value": "vssadmin.exe"},
                {"field": "command_line", "operator": "contains", "value": "shadow delete"}
            ]),
            "mitre_tactic": "Impact",
            "mitre_technique_id": "T1490",
            "description": "Flags attempts to delete volume shadow copies, a common ransomware behavior.",
            "is_active": True,
            "source": "database",
        },
        {
            "id": "RULE-NETWORK-C2",
            "name": "Outbound C2 Communication Pattern",
            "severity": "medium",
            "event_type": "network_connection",
            "conditions": json.dumps([
                {"field": "dst_ip", "operator": "not_equals", "value": "10.0.0.0/8"},
                {"field": "protocol", "operator": "equals", "value": "tcp"}
            ]),
            "mitre_tactic": "Command and Control",
            "mitre_technique_id": "T1071",
            "description": "Looks for suspicious outbound connections outside the internal network segmentation.",
            "is_active": True,
            "source": "database",
        },
    ]


class RuleService:
    def __init__(self, db: AsyncSession, engine: RuleEngine = global_rule_engine):
        self.db = db
        self.engine = engine

    async def get_all_rules(self) -> List[Rule]:
        """Lấy toàn bộ danh sách rules từ Database"""
        stmt = select(Rule).order_by(desc(Rule.created_at))
        result = await self.db.execute(stmt)
        return list(result.scalars().all())

    async def create_rule(self, data: Dict[str, Any]) -> Rule:
        """Thêm Rule mới vào database và Hot-reload Engine"""
        rule_id = data.get("id") or f"RULE-{int(time.time() * 1000)}"

        # Validate JSON conditions
        conditions_str = data.get("conditions", "[]")
        if not isinstance(conditions_str, str):
            conditions_str = json.dumps(conditions_str)
        else:
            try:
                json.loads(conditions_str)
            except Exception:
                raise ValueError("conditions không phải là JSON hợp lệ")

        rule = Rule(
            id=rule_id,
            name=data["name"],
            severity=data.get("severity", "medium"),
            event_type=data.get("event_type", "process_creation"),
            conditions=conditions_str,
            mitre_tactic=data.get("mitre_tactic", ""),
            mitre_technique_id=data.get("mitre_technique_id", ""),
            description=data.get("description", ""),
            is_active=data.get("is_active", True),
            source=data.get("source", "database"),
            created_at=datetime.utcnow(),
            updated_at=datetime.utcnow(),
        )
        self.db.add(rule)
        await self.db.commit()
        await self.db.refresh(rule)

        logger.info(f"[RULE SERVICE] Đã tạo Rule mới '{rule.id}'. Kích hoạt hot-reload...")
        await self.engine.load_all_rules(self.db)
        return rule

    async def update_rule(self, rule_id: str, data: Dict[str, Any]) -> Rule:
        """Cập nhật Rule trong DB và hot-reload"""
        stmt = select(Rule).where(Rule.id == rule_id)
        res = await self.db.execute(stmt)
        rule = res.scalar_one_or_none()

        conditions_str = data.get("conditions")
        if conditions_str is not None:
            if not isinstance(conditions_str, str):
                conditions_str = json.dumps(conditions_str)
            else:
                try:
                    json.loads(conditions_str)
                except Exception:
                    raise ValueError("conditions không phải là JSON hợp lệ")

        if not rule:
            # Tạo mới nếu chưa có
            rule = Rule(
                id=rule_id,
                name=data.get("name", "Unnamed Rule"),
                severity=data.get("severity", "medium"),
                event_type=data.get("event_type", "process_creation"),
                conditions=conditions_str or "[]",
                mitre_tactic=data.get("mitre_tactic", ""),
                mitre_technique_id=data.get("mitre_technique_id", ""),
                description=data.get("description", ""),
                is_active=data.get("is_active", True),
                source=data.get("source", "database"),
                created_at=datetime.utcnow(),
                updated_at=datetime.utcnow(),
            )
            self.db.add(rule)
        else:
            for field in ["name", "severity", "event_type", "mitre_tactic", "mitre_technique_id", "description", "is_active", "source"]:
                if field in data and data[field] is not None:
                    setattr(rule, field, data[field])
            if conditions_str is not None:
                rule.conditions = conditions_str
            rule.updated_at = datetime.utcnow()

        await self.db.commit()
        await self.db.refresh(rule)
        logger.info(f"[RULE SERVICE] Đã cập nhật Rule '{rule.id}'. Kích hoạt hot-reload...")
        await self.engine.load_all_rules(self.db)
        return rule

    async def toggle_rule(self, rule_id: str, is_active: bool) -> bool:
        """Bật/tắt 1 Rule trong DB và hot-reload"""
        stmt = select(Rule).where(Rule.id == rule_id)
        res = await self.db.execute(stmt)
        rule = res.scalar_one_or_none()
        if not rule:
            return False

        rule.is_active = is_active
        rule.updated_at = datetime.utcnow()
        await self.db.commit()
        await self.engine.load_all_rules(self.db)
        return True

    async def delete_rule(self, rule_id: str) -> bool:
        """Xóa 1 Rule khỏi DB và hot-reload"""
        stmt = delete(Rule).where(Rule.id == rule_id)
        res = await self.db.execute(stmt)
        await self.db.commit()
        if (res.rowcount or 0) > 0:
            await self.engine.load_all_rules(self.db)
            return True
        return False
